### PR TITLE
Adding init-upgrade option to terraform

### PR DESCRIPTION
**Pull request by PhillypHenning**
init-upgrade option added to bitops schema + config example

Results of convert-schema.sh without `init-upgrade` flag
`script_options: [  -backend-config="KEY1=foo" -backend-config="KEY2=bar"]`

Results of convert-schema.sh with `init-upgrade` flag
`script_options: [  -backend-config="KEY1=foo" -backend-config="KEY2=bar" --upgrade]`

### CODE REVIEW

**Comment by PhillypHenning** [comment]
I'd like to check where the `--upgrade` flag will be added to. 
It should only be added to the command line when running `terraform init`

**Comment by PhillypHenning** [comment]
Results with `--upgrade` + other cli options
```
BITOPS_CONFIG_COMMAND: [  -backend-config="KEY1=foo" -backend-config="KEY2=bar" --upgrade]
terraform init -input=false --upgrade
```

Results with `--upgrade `+ without other cli options
```
BITOPS_CONFIG_COMMAND: [ --upgrade]
terraform init -input=false --upgrade
```

Results without `--upgrade` + without other cli options
```
BITOPS_CONFIG_COMMAND: []
upgrade not found
terraform init -input=false
```

Results without `--upgrade` + with other cli options
```
BITOPS_CONFIG_COMMAND: [  -backend-config="KEY1=foo" -backend-config="KEY2=bar"]
upgrade not found
terraform init -input=false
```

File: docs/example-config-files/terraform.bitops.config.yaml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 terraform:
     cli: 
       backend-config:
-        - KEY1=foo
-        - KEY2=bar
+        KEY1=foo
```
Comment (mickmcgrath13) [inline docs/example-config-files/terraform.bitops.config.yaml:4]: why this change?

File: scripts/terraform/bitops.schema.yaml [hunk 2]
```diff
@@ -16,6 +16,11 @@ terraform:
           type: parameter-list
           parameter: backend-config
           dash_type: "-"
+        init-upgrade:
+          type: boolean
+          parameter: upgrade
+          dash_type: '--'
```
Comment (mickmcgrath13) [inline scripts/terraform/bitops.schema.yaml:22]: should not be necessary. double dash is default

File: docs/example-config-files/terraform.bitops.config.yaml [hunk 3]
```diff
@@ -1,8 +1,9 @@
...
+        KEY2=bar
+      init-upgrade: true
```
Comment (mickmcgrath13) [inline docs/example-config-files/terraform.bitops.config.yaml:6]: if the `--upgrade` flag is only used for `init` and not `apply`, i'd move this to the `options` section because having it here will also include it in `apply`.

File: scripts/terraform/deploy.sh [hunk 4]
```diff
@@ -43,8 +43,14 @@ ln -s /usr/local/bin/terraform-$TERRAFORM_VERSION /usr/local/bin/terraform
 
 # always init first
 echo "Running terraform init"
-terraform init -input=false
+upgrade=""
+if [[ "$BITOPS_CONFIG_COMMAND" == *"--upgrade"* ]];then
```
Comment (mickmcgrath13) [inline scripts/terraform/deploy.sh:47]: what do those asterisks do here?

Comment (mickmcgrath13) [reply scripts/terraform/deploy.sh:47]: why not just have the schema set an environment variable for the `upgrade` option specifically, and then check for that directly rather than try to parse the whole bitops config command?

**Review by mickmcgrath13 — CHANGES_REQUESTED** [review]

File: docs/example-config-files/terraform.bitops.config.yaml (hunk 1 above)
Comment (PhillypHenning) [reply docs/example-config-files/terraform.bitops.config.yaml]: I've updated this to use;
```
        - KEY1=foo
          KEY2=bar
        - KEY3=baz
```

I just wanted to point out that you can use dash or indent for entries

Comment (mickmcgrath13) [reply docs/example-config-files/terraform.bitops.config.yaml]: each syntax should do something different:
```
backend-config:
  - KEY1=foo
  - KEY2=bar
```
this is a list and is essentially the same as: `backend-config:["KEY1=foo","KEY2=bar"]`

whereas this:
```
backend-config:
  KEY1=foo
  KEY2=bar
```
would probably be invalid syntax.  no dashes implies an object, but object syntax would be:
```
backend-config:
  KEY1: foo
  KEY2: bar
```
which would essentially be: `backend-config: { KEY1: "foo", KEY2: "bar" }`

this syntax:
```
backend-config:
  - KEY1=foo
    KEY2=bar
  - KEY3=baz
```
is also invalid syntax because it's trying to be a list (denoted by the first line's `-`), but the second item doesn't have the dash.  I suspect this would break yaml.

it's _possible_ maybe that this would evaluate to: `backend-config: ["KEY1=fooKEY2=bar", "KEY3=baz"]`, but I doubt it..

I think this should be a list only, like:
```
backend-config:
  - KEY1=foo
  - KEY2=bar
```

Comment (PhillypHenning) [reply docs/example-config-files/terraform.bitops.config.yaml]: I will change this, but I did test locally and it worked
It creates the parameters properly using either dash or indent


**config file**
```
terraform:
    cli: 
      backend-config:
        - KEY1=foo
          KEY2=bar
        - KEY3=baz
      init-upgrade: true
    options:
        command: apply
        version: "0.13.2"
        fetch-kubeconfig: false
```

**Output**
```
script_options: [  -backend-config="KEY1=foo" -backend-config="KEY2=bar" -backend-config="KEY3=baz" --upgrade]
```

Comment (mickmcgrath13) [reply docs/example-config-files/terraform.bitops.config.yaml]: strange that works. I haven't looked at the parsing code, but this is essentially saying:
```
{
  terraform: {
    cli: {
      backend-config: [
          {
              "KEY1=foo",
              "KEY2=bar"
          },{
              "KEY3=baz"
          }
      ]
    }
  }
```

**Review by mickmcgrath13 — CHANGES_REQUESTED** [review]

**Review by mickmcgrath13 — CHANGES_REQUESTED** [review]
Can we also get an update to the options configuration section here:
https://github.com/bitovi/bitops/blob/master/docs/tool-configuration/configuration-terraform.md#options-configuration

**Comment by PhillypHenning** [comment]
Updated the supporting doc

**Review by mickmcgrath13 — APPROVED** [review]
Looks good!  Merge & make a release when you have a minute. 🎉
